NERC-CEH/plankton_ml
# Clustering visualisation - notebook and basic streamlit demo

_Approved but want to add a little more, would like to fix the UI bug in the `k-means` view, and if it's quick also add the UMAP dimensionality reduction [as done for document embeddings](https://github.com/NERC-CEH/discoverability/blob/80e92cfa4395c18c6372e38629c765dba6220b5f/topic_modelling/create_bert_topic_model.py#L11) and look at how that affects DBScan before i give up on it_

* Adds a notebook (flattened with ` jupytext --sync notebooks/*ipynb`) showing K-means clustering of plankton image embeddings and some silhouette analysis which indicates how useful this _might_ be for a simple binary classifier for detritus
* Adds a basic `streamlit` demo showing the results of the different notebook experiments - IMO this was a worthwhile proof of concept in that it could be pointed at any image+embeddings collection without much further work...

## To test

If running from scratch (e.g. you don't already have a `chromadb` instance with an index of image filenames and their embeddings) then

```
export PYTHONPATH=.
python scripts/image_embeddings.py 
streamlit run cyto_ml/visualisation/visualisation_app.py
```

This should bring you to the first page of a two-page app which shows a random plankton image and a set of the closest matches by cosine distance between the embeddings, with a button you can smash to start over and try again from another random plankton image. It should look something like this:

![image](https://github.com/user-attachments/assets/5d56df1c-d3fc-4b03-b426-1fb95dc576b6)

This is also (partly) covered in the README.
